common/kvstore/src/main/java/org/apache/spark/util/kvstore/KVStoreSnapshotter.java

```diff
@@ -0,0 +1,152 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.util.kvstore;
+
+import com.google.common.io.ByteStreams;
+
+import java.io.DataInputStream;
+import java.io.DataOutputStream;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.util.Set;
+
+public class KVStoreSnapshotter {
+  private static final int MARKER_END_OF_TYPE = -2;
+  private static final int MARKER_END_OF_FILE = -1;
+
+  private final KVStoreSerializer serializer;
+
+  public KVStoreSnapshotter(KVStoreSerializer serializer) {
+    this.serializer = serializer;
+  }
+
+  public void dump(KVStore store, File snapshotFile) throws Exception {
```

Comment:
Just want to let you know that SPARK-28867 may also want to dump KVStore to HDFS-support filesystem.

Reply:
Nice catch! Thanks for pointing out. 

Maybe we need to have InputStream/OutputStream (or more specific types) as a parameter instead of File so that it can be worked with any filesystem. kvstore module doesn't have Hadoop dependency so it would be ideal to avoid depending on Hadoop directly.

Reply:
Well, `InputStream/OutputStream` could work around this case. But I'm afraid it brings troublesome usage for the caller, as they needs to prepare their own streams(local or HDFS, in or out). And I believe, currently, KVStore is only designed to persist in filesystems and underlying streams should always be file streams. So, exposing InputStream/OutputStream seems an overkill here.

I'd prefer to use HDFS API to support dump to both local and HDFS-supported filesystem and caller could only pass in a `Path` to indicate where they want to dump (just as previous `File` does). And since we're introducing a **dump** feature here, so I think it would be OK to depend on Hadoop directly now. If we don't depend on Hadoop here, we'll still depend it elsewhere.

WDYT ? @HeartSaVioR

Reply:
> Well, InputStream/OutputStream could work around this case. But I'm afraid it brings troublesome usage for the caller, as they needs to prepare their own streams(local or HDFS, in or out). And I believe, currently, KVStore is only designed to persist in filesystems and underlying streams should always be file streams. So, exposing InputStream/OutputStream seems an overkill here.

Sorry I totally disagree this is an workaround - especially the parameter here depending on is pure Java API, which any Java developers must know and familiar with how to deal with.

https://github.com/apache/spark/blob/c7c6b642dccb3fbd5f82b81cff7f88a776976c7d/core/src/main/scala/org/apache/spark/scheduler/ReplayListenerBus.scala#L52-L59

Just look at ReplayListenerBus. What exactly it requires caller to prepare? InputStream. Does it use something other than file except some case of UT? Well, no. That's consideration on API level to open extension.

I'll leave the decision to the committers.

Even if we turn out to agree directly depend on Hadoop, I would just move this to core module and rewrite this as Scala. Writing this to Java was the intentional effort to put this along with other classes in common/kvstore, and I don't think modules in common depend on hadoop - except network-yarn which even define Hadoop dependency as provided. common modules are trying to avoid depending on Hadoop.

Reply:
That's fine. If we want to follow `ReplayListenerBus` ' s way, I think it would be better if  we could provides read/write helper method later, which just return inputStream/outputStream, similar as `EventLoggingListener.openEventLog()` do.

Reply:
> If we want to follow ReplayListenerBus ' s way, I think it would be better if we could provides read/write helper method later, which just return inputStream/outputStream, similar as EventLoggingListener.openEventLog() do.

That's a good idea. As the PR intends to touch only common-kvstore module, we would be better to deal with this in next PR (FOLLOWUP or include in next work).